app/Jobs/TileSingleImage.php

```diff
@@ -106,6 +116,107 @@ public function uploadToStorage()
         }
     }
 
+    /**
+     * Upload the tiles from temporary local storage to the s3 tiles storage disk.
+     *
+     * @param AwsS3V3Adapter $disk S3 filesystem adapter
+     *
+     */
+    public function uploadToS3Storage($disk)
+    {
+        $iterator = $this->getIterator($this->tempPath);
+        $root = isset($disk->getConfig()['root']) && strlen($disk->getConfig()['root']) ?
+            $disk->getConfig()['root'] . "/" : "";
+
+        $uploads = function ($files) use ($disk, $root) {
+            $client = $this->getClient($disk);
+            $bucket = $this->getBucket($disk);
+            $tmpLength = strlen(config('image.tiles.tmp_dir')) + 1;
+
+            foreach ($files as $file) {
+                $path = substr($file, $tmpLength);
+                $prefix = $path[0] . $path[1] . '/' . $path[2] . $path[3] . "/";
+                // @phpstan-ignore-next-line
+                yield $client->putObjectAsync([
+                    'Bucket' => $bucket,
+                    'Key' => "{$root}{$prefix}{$path}",
+                    'SourceFile' => $file,
+                    // Return with error if file already exist
+                    '@http' => [
+                        'headers' => [
+                            'If-None-Match' => '*'
+                        ]
+                    ]
+                ]);
```

Comment:
I don't get why the path/key is built differently to `uploadToStorage()` here. Doesn't it work the same way?

```php
$prefixLength = strlen($this->tempPath) + 1;
$iterator = $this->getIterator($this->tempPath);
$fragment = fragment_uuid_path($this->image->uuid);

// ...

foreach ($iterator as $pathname => $fileInfo) {
    yield $client->putObjectAsync([
        'Bucket' => $bucket,
        'Key' => "{$prefix}{$fragment}/".substr($pathname, $prefixLength),
        'SourceFile' => $fileInfo,
        // Return with error if file already exist
        '@http' => [
            'headers' => [
                'If-None-Match' => '*'
            ]
        ]
    ]);
}
```

Reply:
I see that your approach has the same result but I still don't see an advantage over the implementation in `uploadToStorage()` (rather the opposite as your `$prefix` is determined anew in each loop whereas `$fragment` is set only once). But I leave this up to you as it's only a minor detail.